apps/devhub/tests/test_views.py

```diff
@@ -2539,9 +2539,9 @@ def xpi(self):
                     'rb')
 
     @mock.patch.object(waffle, 'flag_is_active')
+    @mock.patch('devhub.tasks.validate')
```

Comment:
I have no idea about the `raise AssertionError`. These tests are ancient.

Regarding mocking validate and then calling it manually:

We're testing the behavior of running the validator in the background, and then polling for results. The first time we poll, we need to have no results.

After that, we need the validation job to run, and test that the results are stored as expected, and the API call returns those results.

Since celery tasks are eager in tests, and we don't really have any way to control when they run, we need the code that creates the task to succeed, but not start the task. Once we've verified that polling for a pending task works, we need to actually execute it.

So we need the `validate` mock to delay the execution of the task, and the `run_validator` mock to return the mock results rather than run the validator, when it eventually executes.

There is, unfortunately, a missing piece that I didn't notice, which is testing `validate_.assert_called_once_with(upload)`